www/js/plugin/storage.ts

```diff
@@ -0,0 +1,182 @@
+    return JSON.parse(value);
+  } else {
+    return null;
+  }
+}
+
+/* We redundantly store data in both local and native storage. This function checks
+    both for a value. If a value is present in only one, it copies it to the other and returns it.
+  If a value is present in both, but they are different, it copies the native value to
+    local storage and returns it. */
+function getUnifiedValue(key) {
+  let ls_stored_val = localStorageGet(key);
+  return window['cordova'].plugins.BEMUserCache.getLocalStorage(key, false).then((uc_stored_val) => {
+    logDebug(`for key ${key}, uc_stored_val = ${JSON.stringify(uc_stored_val)},
+                                  ls_stored_val = ${JSON.stringify(ls_stored_val)}.`);
+
+    /* compare stored values by stringified JSON equality, not by == or ===.
+      for objects, == or === only compares the references, not the contents of the objects */
+    if (JSON.stringify(ls_stored_val) == JSON.stringify(uc_stored_val)) {
+      logDebug("local and native values match, already synced");
+      return uc_stored_val;
+    } else {
+      // the values are different
+      if (ls_stored_val == null) {
+        console.assert(uc_stored_val != null, "uc_stored_val should be non-null");
+        logDebug(`for key ${key}, uc_stored_val = ${JSON.stringify(uc_stored_val)},
+                                      ls_stored_val = ${JSON.stringify(ls_stored_val)}.
+                                      copying native ${key} to local...`);
+        localStorageSet(key, uc_stored_val);
+        return uc_stored_val;
+      } else if (uc_stored_val == null) {
+        console.assert(ls_stored_val != null);
+        /*
+         * Backwards compatibility ONLY. Right after the first
+         * update to this version, we may have a local value that
+         * is not a JSON object. In that case, we want to munge it
+         * before storage. Remove this after a few releases.
+         */
+        ls_stored_val = mungeValue(key, ls_stored_val);
+        displayErrorMsg(`Local ${key} found, native ${key} missing, writing ${key} to native`);
+        logDebug(`for key ${key}, uc_stored_val = ${JSON.stringify(uc_stored_val)},
+                                      ls_stored_val = ${JSON.stringify(ls_stored_val)}.
+                                      copying local ${key} to native...`);
+        return window['cordova'].plugins.BEMUserCache.putLocalStorage(key, ls_stored_val).then(() => {
+          // we only return the value after we have finished writing
+          return ls_stored_val;
+        });
+      }
+      console.assert(ls_stored_val != null && uc_stored_val != null,
+        "ls_stored_val =" + JSON.stringify(ls_stored_val) +
+        "uc_stored_val =" + JSON.stringify(uc_stored_val));
+      displayErrorMsg(`Local ${key} found, native ${key} found, but different,
+                           writing ${key} to local`);
+      logDebug(`for key ${key}, uc_stored_val = ${JSON.stringify(uc_stored_val)},
```

Comment:
Ditto wrt displayed error message - useful or just annoying/confusing?

Reply:
I changed it to `logWarn`